lib/srv/db/clickhouse/engine_http.go

```diff
@@ -59,31 +60,44 @@ func (e *Engine) handleHTTPConnection(ctx context.Context, sessionCtx *common.Se
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		query, err := getQuery(req)
-		if err != nil {
+		if err := e.handleRequest(req, sessionCtx, tr); err != nil {
 			return trace.Wrap(err)
 		}
+	}
+}
 
-		queryEvent := common.Query{
-			Query:      query,
-			Parameters: []string{fmt.Sprintf("url=%s", req.URL.String())},
-		}
+func (e *Engine) handleRequest(req *http.Request, sessionCtx *common.Session, tr *http.Transport) error {
+	if req.Body != nil {
+		// we have to close the request body since [http.Server] didn't serve it
+		// up for us.
+		defer req.Body.Close()
+		req.Body = io.NopCloser(io.LimitReader(req.Body, teleport.MaxHTTPRequestSize))
+	}
```

Comment:
The diff looks ugly here, but all I did was move most of the logic out of the above `for` loop and into a function, and added these lines to the top of it so I can conveniently defer closing each request body.